sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -0,0 +1,231 @@
+
+            List<Symbol> outputs = querySpec.outputs();
+            if (querySpec.hasAggregates() == true) {
+                querySpec.hasAggregates(false);
+                querySpec.outputs(splitPoints.toCollect());
+                // TODO: Alter the fields on multiSourceSelect.
+            }
+
+            this.removePostGroupingActionsFromQuerySpec(multiSourceSelect, splitPoints);
+            Plan plan = this.createExecutionPlan(context, multiSourceSelect);
+            this.addGroupProtection(plan, splitPoints, groupKeys);
+            this.addFilterProtection(plan, groupKeys, splitPoints, querySpec);
+            this.addPostAggregationProjection(plan, context, splitPoints, groupKeys, querySpec, outputs);
+
+            return plan;
+        };
+
+        /**
+         * Create and return the current execution plan.
+         */
+        private static Plan createExecutionPlan(ConsumerContext context, MultiSourceSelect multiSourceSelect) {
+            Plan plan = context.plannerContext().planSubRelation(multiSourceSelect, context);
+            boolean executesOnHandler = ExecutionPhases.executesOnHandler(
+                context.plannerContext().handlerNode(),
+                plan.resultDescription().nodeIds()
+            );
+            if (!executesOnHandler) {
+                // TODO: support distributed execution.
+                throw new UnsupportedFeatureException("Distribution execution is not supported");
+            }
+
+            return plan;
+        }
+
+        /**
+         * Adds the post aggregation protection to the given plan.
+         */
+        private static void addPostAggregationProjection(
+            Plan plan,
+            ConsumerContext context,
+            SplitPoints splitPoints,
+            List<Symbol> groupKeys,
+            QuerySpec querySpec,
+            List<Symbol> outputs
+        ) {
+            OrderBy orderBy = querySpec.orderBy().orElse(null);
+            Limits limits = context.plannerContext().getLimits(querySpec);
+            ArrayList list = new ArrayList<>(groupKeys);
+            list.addAll(splitPoints.aggregates());
+            Projection postAggregationProjection = ProjectionBuilder.topNOrEval(
+                list,
+                orderBy,
+                limits.offset(),
+                limits.finalLimit(),
+                outputs
+            );
+            plan.addProjection(postAggregationProjection, null, null, outputs.size(), null);
+        }
+
+        /**
+         * Adds the group protection to the given plan.
+         */
+        private void addGroupProtection(
+            Plan plan,
+            SplitPoints splitPoints,
+            List<Symbol> groupKeys
+        ) {
+            ProjectionBuilder projectionBuilder = new ProjectionBuilder(functions);
+
+            GroupProjection groupProjection = projectionBuilder.groupProjection(
+                splitPoints.leaves(),
```

Comment:
I think this here should use `toCollect` instead of `leaves` - we probably should add a test where there is a scalar function used inside the aggregate function